### ISSUE TITLE
Improve `"any"` cardinality estimation

### ISSUE THREAD

**Issue by coszio** [qdrant/qdrant#2707]
## Current Behavior

Detected by a member of our Discord community
https://discord.com/channels/907569970500743200/1152604870545776670

In short, using the `"any"` filter can be slower than using multiple `"value"` filters together.

Example from the discord conversation:
```
a:
POST /collections/my_collection/points/search
{
  "params": {
    "indexed_only": false
  },
  "filter": {
    "should": [
      {
        "key": "objectId",
        "match": {
          "any": [
            "1000017",
            "1000014",
            "10000275",
            "10000208",
            "10000109",
            "10000156",
            "10000171",
            "10000136",
            "10000015",
            "100003"
          ]
        }
      }
    ]
  },
  "vector": [
    -0.00168334,
    ...
  ],
  "limit": 100
}

b:
POST /collections/my_collection/points/search
{
  "params": {
    "indexed_only": false
  },
  "filter": {
    "should": [
      {
        "key": "objectId",
        "match": {
          "value": "1000017"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "1000014"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "10000275"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "10000208"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "10000109"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "10000156"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "10000171"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "10000136"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "10000015"
        }
      },
      {
        "key": "objectId",
        "match": {
          "value": "100003"
        }
      }
    ]
  },
  "vector": [
    -0.00168334,
    -0.00926589,
    ...
  ],
  "limit": 100
}
```
Reported timing
query a -> 115ms
query b -> 0.5ms

Timing for `exact=true` on query b is also very similar to query a

## Expected Behavior

The two filters should be almost equivalent in performance.

## Possible Solution

Fix cardinality estimation for `"any"` filters.

## Alternative Solution
Translate `"any"` filters to multiple `"value"` filters before execution.

## Context (Environment)

User situation is that he had chunked objects and converted it to vectors, then he would query the points only for a subset of these objects using their objectId

Collection config:
```
{
  "result": {
    "status": "green",
    "optimizer_status": "ok",
    "vectors_count": 1852570,
    "indexed_vectors_count": 1852570,
    "points_count": 1852570,
    "segments_count": 8,
    "config": {
      "params": {
        "vectors": {
          "size": 512,
          "distance": "Euclid"
        },
        "shard_number": 1,
        "replication_factor": 1,
        "write_consistency_factor": 1,
        "on_disk_payload": true
      },
      "hnsw_config": {
        "m": 16,
        "ef_construct": 100,
        "full_scan_threshold": 10000,
        "max_indexing_threads": 0,
        "on_disk": false
      },
      "optimizer_config": {
        "deleted_threshold": 0.2,
        "vacuum_min_vector_number": 1000,
        "default_segment_number": 0,
        "max_segment_size": null,
        "memmap_threshold": null,
        "indexing_threshold": 10,
        "flush_interval_sec": 5,
        "max_optimization_threads": 1
      },
      "wal_config": {
        "wal_capacity_mb": 32,
        "wal_segments_ahead": 0
      },
      "quantization_config": {
        "scalar": {
          "type": "int8",
          "quantile": 0.99,
          "always_ram": true
        }
      }
    },
    "payload_schema": {
      "objectId": {
        "data_type": "keyword",
        "points": 1852570
      }
    }
  },
  "status": "ok",
  "time": 0.008824046
}
```

**Comment by agourlay** [comment]
Just had a quick look, it seems that we forgot to set the primary clause on the estimated cardinalities for the `any` matcher.

I'll check if I can push a quick fix with a test.

**Comment by coszio** [comment]
Solution implemented in #2710, will be available in next release
